postgres_db: add dump and restore support

joshmoore opened the pull request:
Originally filed as https://github.com/ansible/ansible-modules-core/pull/2752 by @autotune. 

---

Ended up just recreating the branch with the specific modified file on there per #2751. The issue template has been re-posted below as well:

Issue Type:

Can you help us out in labelling this by telling us what kind of ticket this this? You can say:
Feature Pull Request

Component Name:
postgresql_db module

Ansible Version: 2.1.0

Ansible Configuration:

Standard config.

Environment:

[autotune@ansibledev ansible-modules-core]$ cat /etc/redhat-release
CentOS Linux release 7.1.1503 (Core)

Summary:

This module adds an import and dump state to the core postgresql_db Module that supports .tar, .gz, .xz, .bz, and a standard .sql file:

```

---
- hosts: local
  tasks:
- name:
  2731_postgresql_db: name=ansitest login_user=postgres state=dump login_host=127.0.0.1 target=~/dev/features/2731/ansitest.sql.xz
```

Steps To Reproduce:

Run the above playbook with a database that exists in postgres called "ansitest" and do a dump. Run the same playbook with "state=import" and change name to an new, empty database, and import.

Expected Results:

Either a new dump created in the form of a compressed file or an imported database.

Actual Results:

See above.

autotuner committed:
Feature #2731: added postgres import and dump

joshmoore committed:
Feature #2731: be more permissive of arguments

```
hacking/test-module -m ./ppostgresql_db.py -a "db=example state=dump target=/tmp/out"`
```

failed previously since host, user, and port were required as keywords
in the pg_dump / pg_import methods.

joshmoore committed:
Feature #2731: fixed doc string for validate-modules

```
$ ansible-validate-modules database/postgresql/
```

now passes.

joshmoore committed:
Feature #2731: disable 'password' for dump/restore

joshmoore commented:
Migrated from https://github.com/ansible/ansible-modules-core/pull/4614

nerzhul commented on lib/ansible/modules/database/postgresql/postgresql_db.py:
```diff
@@ -88,7 +88,18 @@
+      - Back up to a destination, requires target as well. Supported formats are
+        .gz, .tar, .bz, .xz, and .sql
+  restore:
+    version_added: "2.2"
```
please fix version_added to 2.3

gundalow commented on lib/ansible/modules/database/postgresql/postgresql_db.py:
```diff
@@ -88,7 +88,18 @@
       - The database state
```
Please add a comment on the `description` to say dump and import where added in 2.3

Also is it import or restore?

gundalow commented on lib/ansible/modules/database/postgresql/postgresql_db.py:
```diff
@@ -88,7 +88,18 @@
+    description:
+      - Restore from a destination. See backup for supported options
+  target:
+    version_added: "2.2"
```
`"2.3"`

gundalow commented on lib/ansible/modules/database/postgresql/postgresql_db.py:
```diff
@@ -88,7 +88,18 @@
     required: false
     default: present
-    choices: [ "present", "absent" ]
+    choices: [ "present", "absent", "dump", "restore" ]
```
Is it import or restore?

gundalow commented on lib/ansible/modules/database/postgresql/postgresql_db.py:
```diff
@@ -88,7 +88,18 @@
     default: present
-    choices: [ "present", "absent" ]
+    choices: [ "present", "absent", "dump", "restore" ]
+    backup:
```
What are these three lines for?
I don't see any `backup` option, also the formatting is wrong.

gundalow commented on lib/ansible/modules/database/postgresql/postgresql_db.py:
```diff
@@ -88,7 +88,18 @@
+  restore:
+    version_added: "2.2"
+    description:
+      - Restore from a destination. See backup for supported options
```
I don't see a backup option in the code

joshmoore replied:
I'm inclined to think these were originally intended as part of the description. I'll migrate.

gundalow commented on lib/ansible/modules/database/postgresql/postgresql_db.py:
```diff
@@ -88,7 +88,18 @@
+  target:
+    version_added: "2.2"
+    description:
+      - File to back up or restore from.
```
You seem to alternate between the terms backup, import and restore. Picking two will make this easier for people to follow.

joshmoore replied:
Agreed. I didn't even realize the discrepancy when taking over https://github.com/ansible/ansible-modules-core/pull/2752

gundalow reviewed (changes requested):
needs_revision

You seem to alternate between the terms backup, import and restore. Picking two will make this easier for people to follow.

gundalow commented:
@autotune Could you please give this a look. Thanks in advance.

joshmoore committed:
Feature #2731: bump added version to 2.3

joshmoore committed:
Feature #2731: replace db_import with db_restore

joshmoore committed:
Feature #2731: add missing version description

joshmoore committed:
Feature #2731: fix 'state' description

joshmoore commented:
I think the latest push addresses all the comments made. Thanks, all.

gundalow reviewed (approved):
shipit

gundalow commented:
Looks like there are some pep8 tidiups needed (reported by Shippable)

```
2017-01-31 16:45:00 Run command: pep8 --max-line-length 160 --config /dev/null --ignore E123,E124,E127,E128,E201,E202,E203,E211,E221,E222,E225,E226,E227,E228,E231,E241,E251,E261,E262,E265,E266,E301,E302,E303,E402,E502,E713,E731,W391,W503 lib/ansible/modules/database/postgresql/postgresql_db.py
2017-01-31 16:45:00 ERROR: PEP 8: lib/ansible/modules/database/postgresql/postgresql_db.py:262:13: E121 continuation line under-indented for hanging indent (legacy)
2017-01-31 16:45:00 ERROR: PEP 8: lib/ansible/modules/database/postgresql/postgresql_db.py:308:13: E121 continuation line under-indented for hanging indent (legacy)
2017-01-31 16:45:00 PEP 8: There were 3 different legacy issues found (32 total):
2017-01-31 16:45:00 PEP 8: Count Code Message
2017-01-31 16:45:00 PEP 8:    27 E111 indentation is not a multiple of four
2017-01-31 16:45:00 PEP 8:     2 E129 visually indented line with same indent as next logical line
2017-01-31 16:45:00 PEP 8:     3 E501 line too long (NNN > NNN characters)
2017-01-31 16:45:00 ERROR: PEP 8: There are 2 issues which need to be resolved.
```

nerzhul reviewed (commented):
please fix the last pep8 reports

joshmoore committed:
Feature #2731: fix pep8 issues

joshmoore commented:
Rebase removed & pep8 fixed.

joshmoore committed:
Feature #2731: put state documentation in a single string

joshmoore commented:
ready_for_review

joshmoore commented:
ready_for_review

joshmoore commented:
bot_status

joshmoore commented:
(I'm unclear why `needs_revision` was added)

joshmoore commented:
@gundalow @nerzhul, sorry for the ping, but could either of you clarify? Cheers.

nerzhul commented:
@joshmoore CI is happy it's good for me

joshmoore commented:
@gundalow: any thoughts?

gundalow commented:
Maybe add another ready_for_review

After that @nerzhul needs to give change his review

gundalow commented:
Also this is a feature request, We've hit feature freeze on 2.3, so this will have to wait till we've hit RC1 (Friday) before it gets merged into `devel` which will be `2.4` at that point

joshmoore commented:
ready_for_review

joshmoore commented:
@nerzhul : you ok to give a green review based on https://github.com/ansible/ansible/pull/20627#issuecomment-284366643 ? (Think it's a GitHub bug since the commits that you commented on have now been force pushed away)

nerzhul commented:
It's okay for me but your change cannot be merged for 2.3, maybe it should be changed to 2.4 @gundalow ?

gundalow commented on lib/ansible/modules/database/postgresql/postgresql_db.py:
```diff
@@ -59,11 +59,14 @@
-    choices: [ "present", "absent" ]
+    choices: [ "present", "absent", "dump", "restore" ]
+  target:
+    version_added: "2.3"
```
`"2.4"`

joshmoore committed:
Bump added version from 2.3 to 2.4

joshmoore commented:
ready_for_review

mattclay commented:
Closing and re-opening to trigger CI.

joshmoore committed:
Fix pep8 and pylint errors

joshmoore committed:
Attempt yaml formatting of documentation string

joshmoore commented:
ready_for_review

nerzhul commented:
shipit

pchauncey commented:
Might just be me/user error, but I'm getting this on both ansible 2.3 and ansible 2.4: 

`fatal: [localhost]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "db": "foo", 
            "encoding": "", 
            "lc_collate": "", 
            "lc_ctype": "", 
            "login_host": "127.0.0.1", 
            "login_password": "", 
            "login_unix_socket": "", 
            "login_user": "postgres", 
            "name": "foo", 
            "owner": "", 
            "port": 5432, 
            "ssl_mode": "prefer", 
            "ssl_rootcert": null, 
            "state": "dump", 
            "target": "/tmp/foo.sql.gz", 
            "template": ""
        }
    }, 
    "msg": "db_dump() got an unexpected keyword argument 'sslmode'"
}


ansible 2.4.0 (testing_PR20627 b4bffa6d8a) last updated 2017/05/08 10:23:49 (GMT -500)
postgresql-9.5.6-1.fc25.x86_64
`
I've tried with ssl_mode=disable with the same results

joshmoore commented:
I've reproduced in an integration test:

```
TASK [postgresql : Dump the db] ************************************************
fatal: [testhost]: FAILED! => {"changed": false, "failed": true, "msg": "db_dump() got an unexpected keyword argument 'sslmode'"}
	to retry, use: --limit @/root/ansible/test/integration/postgresql-y_LlPW.retry

```

I'll push a fix shortly.

joshmoore committed:
Add integration tests for postgres_db:dump/restore

joshmoore committed:
Update dump/restore logic to support new kw-args

Also attempt to support password; integration tests are
still failing.

joshmoore committed:
Revert to postgres user for dump/restore

Passing PGPASSWORD is not working for subprocesses. For the
moment, reverting to the strategy of failing if login_password
is set and using `postgres` for all testing of dump/restore.

joshmoore committed:
Various cleanups to have tests passing

joshmoore committed:
Working tests for {sql,tar} x {,bz2,gz,xz}

joshmoore committed:
Use pg_user to support FreeBSD

joshmoore commented:
ready_for_review

Note: there are a couple of changes that I could see adding on top of this even for 2.4 but would prefer to get this in first since it's been something of a struggle with the repo migration and both the 2.2 and 2.3 releases. Features to follow include:

 * `-Fd` and `-j $CPUS` to allow saving a database in parallel
 * support for `login_password`

pchauncey commented:
Now it appears that the dump and restore actions aren't honoring login_user:

The "present" action works:
```
ansible localhost -m postgresql_db -a "name=foo login_host=127.0.0.1 login_user=postgres state=present" 
localhost | SUCCESS => {
    "changed": false, 
    "db": "foo"
}
```

But the "dump" action doesn't, and it's trying to use my UID instead of postgres:
```
ansible localhost -m postgresql_db -a "name=foo target=/tmp/foo.sql login_host=127.0.0.1 login_user=postgres state=dump" -vvv
localhost | FAILED! => {
    "changed": false, 
    "cmd": "/usr/bin/psql --dbname=foo < /tmp/foo.sql", 
    "failed": true, 
    "invocation": {
        "module_args": {
            "db": "foo", 
            "encoding": "", 
            "lc_collate": "", 
            "lc_ctype": "", 
            "login_host": "127.0.0.1", 
            "login_password": "", 
            "login_unix_socket": "", 
            "login_user": "postgres", 
            "name": "foo", 
            "owner": "", 
            "port": 5432, 
            "ssl_mode": "prefer", 
            "ssl_rootcert": null, 
            "state": "restore", 
            "target": "/tmp/foo.sql", 
            "template": ""
        }
    }, 
    "msg": "psql: FATAL:  role \"patrick\" does not exist\n", 
    "rc": 2, 
    "stdout": "", 
    "stdout_lines": []
}

```

joshmoore committed:
Revert login_ prefixes and re-enable password support

All `login_` keywords are mapped to their non-prefix versions
so the previous changes were effectively using `postgres` for
all actions. With the proper keywords, PGPASSWORD-passing to
the subprocess is now working.

joshmoore commented:
Thanks for the testing @pchauncey. Good you give https://github.com/ansible/ansible/pull/20627/commits/0939ccab45a96422a5f328110d253ebd4900f22a a try?

pchauncey commented:
Hi @joshmoore, thanks for the quick updates, it is using the postgres user now, but:  Still getting an error here from the same ansible command.  I tried a playbook as well, but both result in the error below:

```
$ ansible localhost -m postgresql_db -a "name=foo target=/tmp/foo.sql login_host=127.0.0.1 login_user=postgres state=dump" -vvv
localhost | FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "db": "foo", 
            "encoding": "", 
            "lc_collate": "", 
            "lc_ctype": "", 
            "login_host": "127.0.0.1", 
            "login_password": "", 
            "login_unix_socket": "", 
            "login_user": "postgres", 
            "name": "foo", 
            "owner": "", 
            "port": 5432, 
            "ssl_mode": "prefer", 
            "ssl_rootcert": null, 
            "state": "dump", 
            "target": "/tmp/foo.sql", 
            "template": ""
        }
    }, 
    "msg": "Database query failed: putenv() argument 2 must be string, not None"
}
$ ansible --version
ansible 2.4.0 (pr20627 f4c49664f0) last updated 2017/05/10 08:17:28 (GMT -500)
```

joshmoore committed:
Optionally add password

environ_update doesn't handle None values in the
dictionary to be added to the environment. Adding
check.

joshmoore commented:
Hi @pchauncey. What this really needs is a number of tests per each combination of user/password, etc. In lieu of that, thanks again for hammering on it!

pchauncey commented:
Here are a couple of quick fixes:

1. In db_dump() Host wasn't being passed to pg_dump if db name was supplied (elif should be if)
2. Quotes around host and port were breaking things for pg_dump if specified
3. In db_restore() cmd was only getting the first flag passed (flag[0])

How do these look to you?

```diff
--- a/lib/ansible/modules/database/postgresql/postgresql_db.py
+++ b/lib/ansible/modules/database/postgresql/postgresql_db.py
@@ -241,12 +241,12 @@ def db_dump(module, target,
     flags = ''
     if db:
         flags += ' {0}'.format(pipes.quote(db))
-    elif host:
-        flags += ' --host={0}'.format(pipes.quote(host))
-        if port:
-            flags += ' --port={0}'.format(pipes.quote(port))
+    if host:
+        flags += ' --host={0}'.format(host)
+    if port:
+        flags += ' --port={0}'.format(port)
     if user:
-        flags += ' --username={0}'.format(pipes.quote(user))
+        flags += ' --username={0}'.format(user)
 
     cmd = module.get_bin_path('pg_dump', True)
     comp_prog_path = None
@@ -311,7 +311,8 @@ def db_restore(module, target,
     elif os.path.splitext(target)[-1] == '.xz':
         comp_prog_path = module.get_bin_path('xzcat', True)
 
-    cmd += flags[0]
+    for flag in flags:
+        cmd += flag
 
     if comp_prog_path:
         p1 = subprocess.Popen([comp_prog_path, target], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```

pchauncey committed:
Quick fixes

joshmoore committed:
Refactor login arguments after fixes from pchauncey

The fixes introduced by pchaunchy pointed to further issues
(like no --dbname on PG<=9.2) with the login parameters. This
refactors them and adds further tests.

Note: this will still not pass integration tests due to a further
      issue with pg_dump as a non-admin user:

      pg_restore: [archiver (db)] Error while PROCESSING TOC:
      pg_restore: [archiver (db)] Error from TOC entry 1925; 0 0 COMMENT EXTENSION plpgsql
      pg_restore: [archiver (db)] could not execute query: ERROR:  must be owner of extension plpgsql

joshmoore committed:
Introduce target_opts for passing limiting dumped/restored schemas

The current integration tests (PG version and template DBs) don't
permit a regular user (`{{ db_user1 }}`) access to plpgsql causing
restores to fail. By adding an option for passing arbitrary args to
pg_dump and pg_restore, testing is made easier. This also paves the
way for `-j` usage, once the PG version is bumped.

joshmoore commented:
Thanks for the fixes, @pchauncey. They certainly pointed to further issues (both in the code and in the tests). I've applied the patch with your info and have pusedh a few commits on top of it. Looking forward to more feedback.

pchauncey reviewed (approved):
shipit

pchauncey commented:
Looks great.  Thanks for all of the additional changes!

joshmoore commented:
Gladly! :) Eventually I could use some info on the minimum supported postgres version and similar or perhaps just on how to scope a feature to a given version in order to introduce the `-j` option. But definitely looking forward to calling this PR done first.

joshmoore commented:
Anything else need from me here?

pchauncey commented:
bot_status

joshmoore commented:
Should I close & re-open or will that put us back at square one?

gundalow commented:
I've clicked "rebuild" on shippable, which should remove the `stale_ci` label.

If people are happy after than I think we can merge.

joshmoore commented:
bot_status

joshmoore commented:
Sorry, @gundalow, but how do I go about making sure everyone is happy?

gundalow commented:
Closing and reopening to trigger CI

gundalow commented:
@matburt @pchaunceyn are you happy with this?

gundalow commented:
I'll merge this next week if I don't hear anything otherwise.

joshmoore commented:
Friendly reminder.

pchauncey commented:
Looks good to me

gundalow commented:
Thanks for the PR and reviews.
As this is new functionality this has been merged into `devel` and will be released in `2.4.0`

autotune commented:
@joshmoore additional thanks from me as well for helping drive this through to completion.
